File: scam_message_analyzer/scoring.py
"""Turn a list of findings into a single traffic-light verdict.

The logic is intentionally simple and explainable: any high-severity signal,
or enough smaller ones stacked together, makes the message red.
"""

from scam_message_analyzer.findings import HIGH

RED = "red"
YELLOW = "yellow"
GREEN = "green"

VERDICT_LABEL = {
    RED: "\U0001F534 LIKELY A SCAM",
    YELLOW: "\U0001F7E1 BE CAREFUL",
    GREEN: "\U0001F7E2 PROBABLY OKAY",
}

_RED_WEIGHT_THRESHOLD = 4


def score(findings):
    """Return ``(verdict, total_weight)`` for the given findings."""

    total = sum(finding.weight for finding in findings)
    has_high = any(finding.severity == HIGH for finding in findings)

    if not findings:
        return GREEN, 0
    if has_high or total >= _RED_WEIGHT_THRESHOLD:
        return RED, total
    return YELLOW, total

File: web/src/scoring.js
// Turn a list of findings into one traffic-light verdict.
// Ported from scam_message_analyzer/scoring.py.

import { HIGH } from "./findings.js";

export const RED = "red";
export const YELLOW = "yellow";
export const GREEN = "green";

export const VERDICT_LABEL = {
  [RED]: "🔴 LIKELY A SCAM",
  [YELLOW]: "🟡 BE CAREFUL",
  [GREEN]: "🟢 PROBABLY OKAY",
};

const RED_WEIGHT_THRESHOLD = 4;

/** Return [verdict, totalWeight] for the given findings. */
export function score(findings) {
  const total = findings.reduce((sum, f) => sum + (f.weight || 0), 0);
  const hasHigh = findings.some((f) => f.severity === HIGH);

  if (findings.length === 0) return [GREEN, 0];
  if (hasHigh || total >= RED_WEIGHT_THRESHOLD) return [RED, total];
  return [YELLOW, total];
}
